### PR TITLE
chore: update lint-related configurations

### PR DESCRIPTION
**Pull request by darcyYe**
#### Intro ####
chore: update lint-related configurations

#### Linear Issue ####
https://linear.app/silverhand/issue/LOG-534/update-lint-related-configurations-silverhandconfigs

#### Test ####
Test of the configuration can be found:

https://github.com/logto-io/js/pull/111
https://github.com/logto-io/logto/pull/155

### CODE REVIEW

File: packages/eslint-config/index.js [hunk 1]
```diff
@@ -319,13 +323,24 @@ module.exports = {
          * Use a more strict `assert` instead.
          */
         'no-restricted-imports': [...xo.rules['no-restricted-imports'], 'assert'],
+        'no-warning-comments': [1, { terms: ['todo', 'fixme'], location: 'start' }],
         // https://github.com/prettier/eslint-config-prettier#curly
         curly: ['error', 'all'],
         /**
-         * eslint-config-xo-typescript disabled this rule by default, need to enable it
+         * Eslint-config-xo-typescript disabled this rule by default, need to enable it
```
Comment (gao-sun) [inline packages/eslint-config/index.js:330]: `aslant`

File: packages/eslint-config/index.js [hunk 2]
```diff
@@ -319,13 +323,24 @@ module.exports = {
          * Use a more strict `assert` instead.
          */
         'no-restricted-imports': [...xo.rules['no-restricted-imports'], 'assert'],
+        'no-warning-comments': [1, { terms: ['todo', 'fixme'], location: 'start' }],
```
Comment (simeng-li) [inline packages/eslint-config/index.js:326]: @gao-sun should we keep the TODO warning lint

File: packages/eslint-config/index.js [hunk 3]
```diff
@@ -222,6 +222,10 @@ const unicornRules = {
   'unicorn/prefer-node-protocol': 'off',
   // Conflict with `@silverhand/fp/no-mutating-methods` when calling `.slice().sort()`
   'unicorn/prefer-spread': 'off',
+  'unicorn/no-await-expression-member': 'off',
```
[the remark this answers is not in the document]
Comment (simeng-li) [reply packages/eslint-config/index.js:225]: should we follow this [rule](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/no-await-expression-member.md
)?
Looks reasonable to me

File: packages/eslint-config/index.js [hunk 4]
```diff
@@ -222,6 +222,10 @@ const unicornRules = {
...
+  'unicorn/template-indent': 'off',
+  'unicorn/prefer-export-from': 'off',
```
Comment (simeng-li) [inline packages/eslint-config/index.js:227]: same here , should we follow the approach?

File: packages/eslint-config/index.js [hunk 5]
```diff
@@ -222,6 +222,10 @@ const unicornRules = {
...
+  'unicorn/prefer-code-point': 'warn',
```
Comment (simeng-li) [inline packages/eslint-config/index.js:228]: https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/prefer-code-point.md
also could update our code?

image

@gao-sun this is the only placing using `charCodeAt`

File: packages/eslint-config/index.js [hunk 6]
```diff
@@ -319,13 +323,24 @@ module.exports = {
...
+        'no-warning-comments': [1, { terms: ['todo', 'fixme'], location: 'start' }],
         // https://github.com/prettier/eslint-config-prettier#curly
         curly: ['error', 'all'],
         /**
-         * eslint-config-xo-typescript disabled this rule by default, need to enable it
+         * Eslint-config-xo-typescript disabled this rule by default, need to enable it
          * https://github.com/xojs/eslint-config-xo-typescript/blob/main/index.js#L446
          */
         '@typescript-eslint/no-non-null-assertion': 'error',
+        '@typescript-eslint/member-ordering': [
```
Comment (simeng-li) [inline packages/eslint-config/index.js:334]: which case are we intent to modify the default rule settings?

File: packages/eslint-config/index.js (hunk 5 above)
Comment (darcyYe) [reply packages/eslint-config/index.js:228]: I tried to replace .charCodeAt() with .codePointAt(), but the return type of charCodeAt() is number or NaN while the return type of charCodeAt() is number or undefined. Hence the lint-rule-checker throw error that operand '+' can not work between a number type variable (accumulator) and undefined variable even if I did type-check. Any idea that I can follow to walk around?

Comment (simeng-li) [reply packages/eslint-config/index.js:228]: cool, let me double check

File: packages/eslint-config/index.js (hunk 4 above)
Comment (darcyYe) [reply packages/eslint-config/index.js:227]: logto/js/packages/client/src/index.ts
image
`IDToken` is exported but not used. When we try to use `IDToken` in some other files, should we just export `IDToken` from where it was defined?

File: packages/eslint-config/index.js (hunk 5 above)
Comment (simeng-li) [reply packages/eslint-config/index.js:228]: image
maybe this approach?

File: packages/eslint-config/index.js [hunk 7]
```diff
@@ -222,6 +222,10 @@ const unicornRules = {
   'unicorn/prefer-node-protocol': 'off',
   // Conflict with `@silverhand/fp/no-mutating-methods` when calling `.slice().sort()`
   'unicorn/prefer-spread': 'off',
+  'unicorn/no-await-expression-member': 'off',
+  'unicorn/template-indent': 'off',
```
Comment (darcyYe) [inline packages/eslint-config/index.js:226]: fixed

File: packages/eslint-config/index.js (hunk 6 above)
Comment (darcyYe) [reply packages/eslint-config/index.js:334]: logto/js/packages/client/src/index.ts
image

This rule is added to ensure the order when we try to define a class and all its members (including constructor, fields and methods).

File: packages/eslint-config/index.js (hunk 4 above)
Comment (simeng-li) [reply packages/eslint-config/index.js:227]: o it is being used by the react SDK. 

image

Will do

File: packages/eslint-config/index.js [hunk 8]
```diff
@@ -249,27 +249,28 @@ const fpRules = {
...
+//       spaces: 2,
+//     },
+//   ],
+//   'sql/no-unsafe-query': [
+//     1, // Warn use of SQL inside of template literals without the sql tag
+//     {
+//       allowLiteral: false,
+//     },
+//   ],
+// };
```
Comment (gao-sun) [inline packages/eslint-config/index.js:273]: or we can just delete the plugin if `unicorn/template-indent` works great?

File: packages/eslint-config/index.js [hunk 9]
```diff
@@ -322,7 +323,7 @@ module.exports = {
         // https://github.com/prettier/eslint-config-prettier#curly
         curly: ['error', 'all'],
         /**
-         * eslint-config-xo-typescript disabled this rule by default, need to enable it
+         * Pack eslint-config-xo-typescript disabled this rule by default, need to enable it
```
Comment (gao-sun) [inline packages/eslint-config/index.js:326]: "Pack" for?

Comment (darcyYe) [reply packages/eslint-config/index.js:326]: comments need to be capitalized lol

File: packages/eslint-config/index.js (hunk 8 above)
Comment (darcyYe) [reply packages/eslint-config/index.js:273]: Make sense. But what was the reason for introducing this rule?

File: packages/eslint-config/index.js (hunk 9 above)
Comment (gao-sun) [reply packages/eslint-config/index.js:326]:
```suggestion
         * `eslint-config-xo-typescript` disabled this rule by default, need to enable it
```

File: packages/eslint-config/index.js [hunk 10]
```diff
@@ -251,20 +251,8 @@ const fpRules = {
...
-      ignoreTagless: false,
-      ignoreStartWithNewLine: true,
-    },
-    {
-      spaces: 2,
-    },
-  ],
   'sql/no-unsafe-query': [
-    2,
+    1, // Warn use of SQL inside of template literals without the sql tag
```
Comment (gao-sun) [inline packages/eslint-config/index.js:255]:
```suggestion
    'error', // Warn use of SQL inside of template literals without the sql tag
```

**Review by gao-sun — APPROVED** [review]

**Pull request merged**